### PR TITLE
TF counter in the DataHeader

### PR DESCRIPTION
**Pull request by shahor02**
* Add ever-incrementing TF counter to DataHeader
The DataHeader::firstTForbit will not be unique when the same TFs are replayed in loop, this counter will resolve the ambiguity.

* Set the tfCounter in RawDataReader and CTFReader 

@ironMann this would need to be set also in the STFbuilder

### CODE REVIEW

File: DataFormats/Headers/include/Headers/DataHeader.h [hunk 1]
```diff
@@ -666,6 +666,7 @@ struct DataHeader : public BaseHeader {
   using SplitPayloadPartsType = uint32_t;
   using PayloadSizeType = uint64_t;
   using TForbitType = uint32_t;
+  using TFCounterType = uint32_t;
 
   //static data for this header type/version
   static constexpr uint32_t sVersion{2};
```
Comment (ironMann) [inline DataFormats/Headers/include/Headers/DataHeader.h:672]: Here we go again. :)  Please bump the version

File: DataFormats/Headers/include/Headers/DataHeader.h [hunk 2]
```diff
@@ -711,10 +712,15 @@ struct DataHeader : public BaseHeader {
   //___NEW STUFF GOES BELOW
 
   ///
-  /// first orbit of time frame as unique identifier within the run
+  /// first orbit of time frame as unique identifier within the run (unless the TF is replayed)
   ///
   TForbitType firstTForbit;
 
+  ///
```
Comment (ironMann) [inline DataFormats/Headers/include/Headers/DataHeader.h:719]: can you please add a comment here stating this was added in Version3. I need to convert old recorded data to new DH versions, and looking this up via git is not ideal. (and we don't keep the old structure definitions around, like for the raw header) Thanks

**Comment by mkrzewic** [comment]
This is quite a narrow use case (testing/debugging), adding a member to cover this scenario seems to me a bit excessive? Is there no other way we can make sure the ids are unique? I'm thinking more along the lines of having the replay software faking an incrementing orbit counter and setting it in the header of the to be replayed TF.

**Comment by davidrohr** [comment]
@mkrzewic : The problem is that this can break the reconstruction, since in that case you don't have the correct first orbit. I think it is a good idea to disentangle the tf counter used for data distribution, and the physical orbit, which might repeat in data replay cases.

**Comment by ironMann** [comment]
DataDistribution can live with one copy of this information. As we discussed this morning, there would be a special MetaData in the TF (one o2 message) that carries the tfid. o2-raw-proxy would read this, and use it to route and aggregate data while i's flowing though the DPL. On output, DPL-> StfSender for example, we revert to the original metadata message as received on the input.

**Comment by shahor02** [comment]
@mkrzewic The TF replay will our main use case for next 1.5 years. 
Faking the firstTForbit in the header is what is done now, but this makes it inconsistent with the interaction records contained in the message payload.
Note that this counter does not affect the size of the DataHeader.

**Comment by mkrzewic** [comment]
@shahor02 but we don't need this information in every message. what about @ironMann comment? A single descriptor valid for the whole collection of messages (TF, or part of it), no pollution in the data model, no unused field in the oodles of data we ARE going to store after the first 1.5 years. Once adopted (although envisioned from the very beginning) the  concept is much more elastic - you get to add whatever metadata you like without having to touch the main headers again.

**Comment by shahor02** [comment]
@mkrzewic As far as I understand, the comment of @ironMann is relevant for DD only, which does not need the DataHeader at all.
In opposite, the DPL will need a unique TF ID in every message to disentangle messages from different TFs, for the reasons discussed when introducing the firstTForbit. Since the latter is not unique in case of the replay, we have 2 choices:

1) firstTForbit is hacked in case of the replay, so that it stays unique and otherwise meaningless (i.e. TF counter).
But some devices need the true 1st TF orbit to process their data, so we should introduce a special O2 message e.g. TFHeader injected into DPL for every TF (also for digits based MC reconstruction)

2) extend real (but ambiguous in case of replay) firstTForbit by the an extra TF counter (you can consider firstTForbit+tfCounter as a single 64bit identifier you wanted to add originally).

This was the content of message I sent >2weeks age and did not get any response.

**Comment by ironMann** [comment]
@shahor02 DataDistribution attaches DataHeader to every piece of data it has. But since I always move data in complete (S)TFs, I don't actually need to replicate TFid to every and each DataHeader. that's all I wanted to say.

I was under impression that DPL will use the DataProcessingHeader to attach this info to all O2Message it gets from me?

**Comment by shahor02** [comment]
@ironMann I meant you are adding the DataHeader just to comply with O2 requirements, you don't use it.
The idea was to eventually drop the DataProcessingHeader. If the DPL wants to keep it and rely on it to disentangle between TFs, fine for me

**Comment by davidrohr** [comment]
Just to add my opinion: i think we need this distinction between firstOrbit and timeframeId. No way to avoid that. I like to keep the dataheader clean, but i also like dropping the additional dataprocessing header in case it is obsolete. But ok, we have to store the information somewhere, so in the end i think both, either adding the field as Ruben did, or having an extra header with such information is ok.

**Comment by ironMann** [comment]
So to recap: I'll have data coming from the readout where they provide the tfID (a). And I'll have data from sim (b), where we can start counting tfid from zero up. 
The question is what to do when we loop over the same data set of TFs. If I repeatedly inject the same tf into DPL, should I fake that tfid to be something new? Does the reco need the 'real' tfid in (a)?

**Comment by matthiasrichter** [comment]
I would refrain from adding this additional member to DataHeader.

With the firstTForbit member we already have a member to uniquely relate data to time. What we need to do in the case of data replay is to generate unique time slices in the DPL for the TFs with the same orbit information. I need to check how the time slice information is generated from the firstTFmember, or @ktf can comment.

I think we originally considered dropping the DPL DataProcessingHeader once we have some unique time information in DataHeader. But now we see that this information is coupled to data (so having it in DataHeader makes sense), while we want to process one or multiple TFs in a data replay configuration, and this means we need to create a new DPL time slice for every of those TFs, and DPL does the routing based on the time slice information.

**Comment by shahor02** [comment]
> The question is what to do when we loop over the same data set of TFs. If I repeatedly inject the same tf into DPL, should I fake that tfid to be something new? Does the reco need the 'real' tfid in (a)?

Data processing needs "real" firstTForbit, in opposite, the TF counter can be anything, provided it is unique. So, in the 2nd commit of this PR I simply increment the tfCounter each time the RawReader of CTFReader injects new TF into the DPL, also in case of replaying the same TF. 

@matthiasrichter could you present more arguments why the firstTForbit (which may pose problems in case of replay, cosmic, calibration data etc) should be placed in the DataHeader while the counter which would always disentangle different processing periods should not be there ?

Yes, the same info could be played by the timecounter in the DataProcessingHeader, but then I don't see much sense in having 2 separate headers which always should come together.

Please note that this topic is currently a blocker for multi-TF processing, for me any solution is ok provided we come to it quickly.

**Comment by davidrohr** [comment]
This is also a blocker for the full system test with more than 1 GPU, so I think whatever solution we go for, we should have it soon.
In principle, I'd even say LGTM and we can revert and clean up later, but if @ironMann says he needs to reformat some old data if we change the data header, that might be a bit annoying.
Just for the records: I believe if this data header is stored with the data, having the DPL timeframe id in a separate header is the cleaner solution, since this information is only needed for routing and doesn't belong to the data, but I wouldn't want to wait for it too long.

**Comment by ironMann** [comment]
I'm leaning towards what @matthiasrichter suggested (what we discussed today). David, I can quickly add the same increment of the first orbit as the @shahor02. Until this is 'properly' resolved... Would that help short-term?

**Comment by davidrohr** [comment]
@ironMann : Yes, that would help me for now.

**Comment by shahor02** [comment]
@ironMann can I know what you refer to?

**Comment by ironMann** [comment]
@shahor02 falsifying the firstTfOrbit for repeated TFs

**Comment by shahor02** [comment]
@ironMann But avoiding the discrepancy between the DataHeader::firstTFOrbit and orbits stored in the message payload was the whole point of this PR. With such a discrepancy the firstTFOrbit becomes just a dummy counter against which you are protesting.
And detectors which need the true firstTFOrbit to process their data will need to ignore this info in the header and recalculate it from the orbits stored in the data (using ad hoc knowledge of run 1st orbit and NHB per TF), which I would prefer to avoid.

**Comment by ironMann** [comment]
@shahor02 I understand the issue. The ugly thruth is that today two groups of us started working on two different solutions for this. I propse that @ktf and @mkrzewic come up with a decision where to keep the tfid: in the DataProcessingHeader (can this actually be retired?) or in the DataHeader.

**Comment by ktf** [comment]
Since DataProcessingHeader exists solely because there is years long resistance in integrating Data Processing needs in DataHeader, I am happy to drop it if the following requirements are satisfied and provided from upstream:

1) We have a way to determine a time to do a time query to CCDB, without having to inspect the payloads. The TFOrbit (+ some extra lookup) satisfies that.
2) The timeframe parts to be processed at the same time are uniquely identified without having to inspect the payloads. I guess tfOrbit + Ruben's counter satisfies that.

I actually found this proposal from 2018 https://indico.cern.ch/event/782003/contributions/3258880/attachments/1774424/2884541/2018-dec-wp1.pdf where "iteration" would have covered perfectly this case. 

My point is always the same: the overhead of having an extra header is larger than what the requested modification would require (especially since in any case we would need the extra padding).

**Comment by shahor02** [comment]
> @ktf    1. We have a way to determine a time to do a time query to CCDB, without having to inspect the payloads. The TFOrbit (+ some extra lookup) satisfies that.

to be precise: to go the absolute time for the CCDB query on top of the firstTForbit we need either runID or any absolute time-stamp within the run (not necessarily within the given TF)

**Comment by ktf** [comment]
@shahor02 yes, that's what I meant. ;-)

**Comment by shahor02** [comment]
then shall we define such a time stamp or run number now, to avoid another round in a few weeks?

**Comment by mkrzewic** [comment]
@shahor02 experience teaches there definitely WILL be another round, as is this discussion:) Please prepare a proposal for next week wp1 meeting, we'll discuss it.
Let me repeat my preference: DataHeader should only contain things that are strictly necessary for every single message, anything that describes a bunch of them (like a sub- TF or a group of data going along some edge in a topology should be described with a *part* that tags along, a part that we can modify at will without breaking the interfaces - these things like to change quickly and we need to keep the headers stable.
I have in principle very little against making the tfid 64bit instead of 32, but please don't let it grow beyond that. we really don't need that big a number.

**Comment by shahor02** [comment]
@mkrzewic I don't know which part of current 88 byte-long DataHeader matches your preferences, but as I wrote, I don't care where you put this extra info, I just need it quick as it is a blocker since a few weeks. 
If you prefer, move it to DPH, though I don't understand the logic of putting firstTForbit to the DataHeader, and augmenting it by another 56B long DPH with complementary info to make it usable.
The same goes for the additional timestamp or run-number mentioned above. If you want to have it in any of DPL headers (and I guess at least Giulio wants it there, in order to use for the CCDB query), you decide where to put it, I don't understand what extra proposal you want to hear at the meeting.

My personal preference would be to not have at all this info in the DataHeader.  I would publish instead a TF header O2 message to which any device can subscribe. This was in the email message I sent 2 weeks ago to all participants of this thread and did not get any answer. To be viable solution, such header would need to be published also by the DataDistribution.

**Comment by ktf** [comment]
@shahor02 at this point, that's also fine with me, if the same message comes also from data distribution, and not only from your reader.

**Comment by shahor02** [comment]
@ktf, sorry, did not understand what you mean. If you refer to alternative TF header message, it should be sent by whoever injects the data to DPL and the MC workflows should should be preceded by a special workflow that fakes this message.

**Comment by ironMann** [comment]
> My personal preference would be to not have at all this info in the DataHeader. I would publish instead a TF header O2 message to which any device can subscribe. This was in the email message I sent 2 weeks ago to all participants of this thread and did not get any answer. To be viable solution, such header would need to be published also by the DataDistribution.

This is exactly what @mkrzewic , @matthiasrichter and I are suggesting. This TF-info message can be extended at any time then (I would also revert the firstOrbit and put it in there). DPL can include a copy of the record with each input and output data block.

**Comment by ktf** [comment]
@shahor02 yes, that's what I was referring to. Let's move the whole logic inside some custom agreed payload and that's it.

**Comment by ktf** [comment]
@ironMann that's fine.

**Comment by shahor02** [comment]
Failure are unrelated, merging.

**Comment by mkrzewic** [comment]
given the conclusion above I imagine this was merged by mistake?

**Comment by shahor02** [comment]
yes, sorry, too many windows open...

**Comment by ironMann** [comment]
#3912 for the proposed solution

**Pull request merged**